### ISSUE TITLE
Random crashes on El Captain

### ISSUE THREAD

**Issue by sbktrifork** [iscsi-osx/iSCSIInitiator#83]
Hi 

I'm using El Captain with an iscsi target on Linux CentOS 7 (tgtd 1.0.55). 

When using the iscsi disk, El Captain crashes the hard way after a while (minutes/hours):

Anonymous UUID:       638A8723-0F3E-D364-06B6-191E59108B2E

Sun Feb 19 22:40:57 2017

*** Panic Report ***
panic(cpu 0 caller 0xffffff80289cc7ba): "Double fault at 0xffffff7fa91792c7, registers:\n" "CR0: 0x000000008001003b, CR2: 0xffffff91325ebe18, CR3: 0x000000002bcef000, CR4: 0x00000000000626e0\n" "RAX: 0x0000000000ff006b, RBX: 0xffffff91335dbe80, RCX: 0xffffff91325ebe20, RDX: 0xffffff803f264200\n" "RSP: 0xffffff91325ebe20, RBP: 0xffffff91335dbec0, RSI: 0xffffff803fbf4d00, RDI: 0xffffff803c640600\n" "R8:  0x0000000000ff005c, R9:  0x0000000000000000, R10: 0xffffff803fc14cd8, R11: 0x000000000002aa50\n" "R12: 0xbd9e22c037f000dd, R13: 0xffffff803c640600, R14: 0xffffff803f264200, R15: 0xffffff803fbf4d00\n" "RFL: 0x0000000000010282, RIP: 0xffffff7fa91792c7, CS:  0x0000000000000008, SS:  0x0000000000000000\n" "Error code: 0x0000000000000000\n"@/Library/Caches/com.apple.xbs/Sources/xnu/xnu-3248.60.11.2.1/osfmk/i386/trap_native.c:168
Backtrace (CPU 0), Frame : Return Address
0xffffff810a128cf0 : 0xffffff80288d7b92 
0xffffff810a128d70 : 0xffffff80289cc7ba 
0xffffff810a128ed0 : 0xffffff80289ea45f 
0xffffff91335dbec0 : 0xffffff7fa917899e 
0xffffff91335dbf20 : 0xffffff7fa91767c3 
0xffffff91335dbf40 : 0xffffff8028eafff1 
0xffffff91335dbf80 : 0xffffff8028eb00e6 
0xffffff91335dbfb0 : 0xffffff80289c6537 
      Kernel Extensions in backtrace:
         com.github.iscsi-osx.iSCSIInitiator(1.0)[BF36C32C-432C-3879-8338-77086886983E]@0xffffff7fa9174000->0xffffff7fa917efff
            dependency: com.apple.iokit.IOSCSIParallelFamily(3.0.0)[7C94C05D-4761-3D03-AB56-4CB7FCFA6039]@0xffffff7fa911a000

BSD process name corresponding to current thread: kernel_task

Mac OS version:
15G1217

Kernel version:
Darwin Kernel Version 15.6.0: Mon Jan  9 23:07:29 PST 2017; root:xnu-3248.60.11.2.1~1/RELEASE_X86_64
Kernel UUID: 0BADD3CA-3553-380F-8A09-3B02BA961B45
Kernel slide:     0x0000000028600000
Kernel text base: 0xffffff8028800000
__HIB  text base: 0xffffff8028700000
System model name: Macmini5,1 (Mac-8ED6AF5B48C039E1)

System uptime in nanoseconds: 2156379421395
last loaded kext at 1069796035853: com.apple.filesystems.afpfs	11.0 (addr 0xffffff7fab522000, size 368640)
loaded kexts:
com.github.iscsi-osx.iSCSIInitiator	1
com.kernsafe.driver.kscsiprt	1.0.0d1
com.apple.filesystems.afpfs	11.0
com.apple.nke.asp-tcp	8.0.0
com.apple.filesystems.msdosfs	1.10
com.apple.driver.AppleBluetoothMultitouch	90.3
com.apple.iokit.SCSITaskUserClient	3.7.7
com.apple.nke.l2tp	1.9
com.apple.nke.pptp	1.9
com.apple.driver.AudioAUUC	1.70
com.apple.driver.AppleHWSensor	1.9.5d0
com.apple.driver.AppleTyMCEDriver	1.0.2d2
com.apple.driver.AGPM	110.22.0
com.apple.driver.ApplePlatformEnabler	2.6.0d0
com.apple.filesystems.autofs	3.0
com.apple.driver.AppleOSXWatchdog	1
com.apple.driver.AppleMikeyHIDDriver	124
com.apple.driver.AppleHDAHardwareConfigDriver	274.12
com.apple.driver.AppleHDA	274.12
com.apple.driver.AppleMikeyDriver	274.12
com.apple.driver.ApplePolicyControl	3.12.9
com.apple.driver.AppleUpstreamUserClient	3.6.1
com.apple.driver.AppleMCCSControl	1.2.13
com.apple.driver.pmtelemetry	1
com.apple.iokit.IOUserEthernet	1.0.1
com.apple.iokit.IOBluetoothSerialManager	4.4.6f1
com.apple.Dont_Steal_Mac_OS_X	7.0.0
com.apple.driver.AppleHV	1
com.apple.iokit.BroadcomBluetoothHostControllerUSBTransport	4.4.6f1
com.apple.driver.AppleSMCPDRC	1.0.0
com.apple.driver.AppleIntelHD3000Graphics	10.0.0
com.apple.driver.AppleThunderboltIP	3.0.8
com.apple.driver.AppleLPC	3.1
com.apple.driver.AppleMuxControl	3.12.9
com.apple.driver.ACPI_SMC_PlatformPlugin	1.0.0
com.apple.driver.AppleIntelSlowAdaptiveClocking	4.0.0
com.apple.driver.AppleFIVRDriver	4.1.0
com.apple.driver.AppleIntelSNBGraphicsFB	10.0.0
com.apple.iokit.IOBluetoothUSBDFU	4.4.6f1
com.apple.driver.AppleIRController	327.6
com.apple.driver.AppleFileSystemDriver	3.0.1
com.apple.AppleFSCompression.AppleFSCompressionTypeDataless	1.0.0d1
com.apple.AppleFSCompression.AppleFSCompressionTypeZlib	1.0.0
com.apple.BootCache	38
com.apple.iokit.IOAHCIBlockStorage	2.8.5
com.apple.driver.AppleFWOHCI	5.5.4
com.apple.iokit.AppleBCM5701Ethernet	10.2.0
com.apple.driver.AppleSDXC	1.7.0
com.apple.driver.AirPort.Brcm4331	800.20.24
com.apple.driver.AirPort.Brcm4360	1040.1.1a6
com.apple.driver.AppleAHCIPort	3.1.8
com.apple.driver.usb.AppleUSBUHCIPCI	1.0.1
com.apple.driver.usb.AppleUSBEHCIPCI	1.0.1
com.apple.driver.AppleRTC	2.0
com.apple.driver.AppleACPIButtons	4.0
com.apple.driver.AppleHPET	1.8
com.apple.driver.AppleSMBIOS	2.1
com.apple.driver.AppleACPIEC	4.0
com.apple.driver.AppleAPIC	1.7
com.apple.driver.AppleIntelCPUPowerManagementClient	218.0.0
com.apple.nke.applicationfirewall	163
com.apple.security.quarantine	3
com.apple.security.TMSafetyNet	8
com.apple.driver.AppleIntelCPUPowerManagement	218.0.0
com.apple.security.SecureRemotePassword	1.0
com.apple.driver.AppleBluetoothHIDKeyboard	181
com.apple.driver.AppleHIDKeyboard	181
com.apple.driver.IOBluetoothHIDDriver	4.4.6f1
com.apple.driver.AppleMultitouchDriver	304.12
com.apple.driver.AppleHIDTransport	5
com.apple.iokit.IOSCSIBlockCommandsDevice	3.7.7
com.apple.nke.ppp	1.9
com.apple.AppleGraphicsDeviceControl	3.12.9
com.apple.kext.triggers	1.0
com.apple.driver.DspFuncLib	274.12
com.apple.kext.OSvKernDSPLib	525
com.apple.iokit.IOSCSIParallelFamily	3.0.0
com.apple.iokit.IOSurface	108.2.3
com.apple.iokit.IOSerialFamily	11
com.apple.driver.AppleSMBusController	1.0.14d1
com.apple.iokit.IOFireWireIP	2.2.6
com.apple.driver.CoreCaptureResponder	1
com.apple.iokit.IOBluetoothHostControllerUSBTransport	4.4.6f1
com.apple.iokit.IOBluetoothFamily	4.4.6f1
com.apple.driver.AppleHDAController	274.12
com.apple.iokit.IOHDAFamily	274.12
com.apple.iokit.IOAudioFamily	204.4
com.apple.vecLib.kext	1.2.0
com.apple.driver.AppleSMBusPCI	1.0.14d1
com.apple.driver.AppleThunderboltEDMSink	4.1.1
com.apple.driver.AppleBacklightExpert	1.1.0
com.apple.driver.AppleGraphicsControl	3.12.9
com.apple.driver.IOPlatformPluginLegacy	1.0.0
com.apple.driver.IOPlatformPluginFamily	6.0.0d7
com.apple.iokit.IOSlowAdaptiveClockingFamily	1.0.0
com.apple.driver.AppleSMC	3.1.9
com.apple.iokit.IONDRVSupport	2.4.1
com.apple.iokit.IOGraphicsFamily	2.4.1
com.apple.iokit.IOSCSIArchitectureModelFamily	3.7.7
com.apple.driver.usb.IOUSBHostHIDDevice	1.0.1
com.apple.iokit.IOUSBHIDDriver	900.4.1
com.apple.driver.usb.cdc	5.0.0
com.apple.driver.usb.networking	5.0.0
com.apple.driver.usb.AppleUSBHostCompositeDevice	1.0.1
com.apple.driver.usb.AppleUSBHub	1.0.1
com.apple.driver.AppleThunderboltDPInAdapter	4.1.3
com.apple.driver.AppleThunderboltDPOutAdapter	4.1.3
com.apple.driver.AppleThunderboltDPAdapterFamily	4.1.3
com.apple.driver.AppleThunderboltPCIDownAdapter	2.0.2
com.apple.driver.AppleXsanScheme	3
com.apple.iokit.IOFireWireFamily	4.6.1
com.apple.iokit.IOEthernetAVBController	1.0.3b3
com.apple.driver.AppleUSBMergeNub	900.4.1
com.apple.driver.AppleThunderboltNHI	4.0.4
com.apple.iokit.IOThunderboltFamily	6.0.2
com.apple.iokit.IO80211Family	1110.26
com.apple.driver.mDNSOffloadUserClient	1.0.1b8
com.apple.iokit.IONetworkingFamily	3.2
com.apple.driver.corecapture	1.0.4
com.apple.iokit.IOAHCIFamily	2.8.1
com.apple.driver.usb.AppleUSBUHCI	1.0.1
com.apple.iokit.IOUSBFamily	900.4.1
com.apple.driver.usb.AppleUSBEHCI	1.0.1
com.apple.iokit.IOUSBHostFamily	1.0.1
com.apple.driver.AppleUSBHostMergeProperties	1.0.1
com.apple.driver.AppleEFINVRAM	2.0
com.apple.driver.AppleEFIRuntime	2.0
com.apple.iokit.IOHIDFamily	2.0.0
com.apple.iokit.IOSMBusFamily	1.1
com.apple.security.sandbox	300.0
com.apple.kext.AppleMatch	1.0.0d1
com.apple.driver.AppleKeyStore	2
com.apple.driver.AppleMobileFileIntegrity	1.0.5
com.apple.driver.AppleCredentialManager	1.0
com.apple.driver.DiskImages	417.4
com.apple.iokit.IOStorageFamily	2.1
com.apple.iokit.IOReportFamily	31
com.apple.driver.AppleFDEKeyStore	28.30
com.apple.driver.AppleACPIPlatform	4.0
com.apple.iokit.IOPCIFamily	2.9
com.apple.iokit.IOACPIFamily	1.4
com.apple.kec.Libm	1
com.apple.kec.pthread	1
com.apple.kec.corecrypto	1.0
Model: Macmini5,1, BootROM MM51.0077.B15, 2 processors, Intel Core i5, 2,3 GHz, 8 GB, SMC 1.76f0
Graphics: Intel HD Graphics 3000, Intel HD Graphics 3000, Built-In
Memory Module: BANK 0/DIMM0, 4 GB, DDR3, 1333 MHz, 0x0198, 0x393930353432382D3035312E4130304C4620
Memory Module: BANK 1/DIMM0, 4 GB, DDR3, 1333 MHz, 0x0198, 0x393930353432382D3035312E4130304C4620
AirPort: spairport_wireless_card_type_airport_extreme (0x14E4, 0xE4), Broadcom BCM43xx 1.0 (5.106.98.100.24)
Bluetooth: Version 4.4.6f1 17910, 3 services, 27 devices, 1 incoming serial ports
Network Service: Ethernet, Ethernet, en0
Network Service: Wi-Fi, AirPort, en1
Serial ATA Device: Samsung SSD 840 EVO 120GB, 120,03 GB
SCSI Device: SCSI Target Device @ 0
USB Device: USB 2.0 Bus
USB Device: Hub
USB Device: BRCM20702 Hub
USB Device: Bluetooth USB Host Controller
USB Device: USB 2.0 Bus
USB Device: Hub
USB Device: IR Receiver
Thunderbolt Bus: Mac mini, Apple Inc., 25.1

**Comment by nsinenian** [comment]
Are you using BETA-5? Also, can you tell me a bit more about the usage
pattern -- are you reading or writing when this happens, or does it not
matter? Is it reproducible?

**Comment by sorenk** [comment]
Sorry for the missing informations. 

Yes, I’m using Beta-5. 
I can’t tell much about the usage pattern or how to reproduce just yet. 
I’m having an iTunes Library on it and a Photos Library. 
Which means that as soon as I mount the disk, OS X photos cloud starts to update photos on the disk. 

What I can tell for now is when I’m not logging into the disk, OS X stays stable. 
I will try different stuff and update the bug report as soon as I have more information. 
The OS X installation is pretty straight forward, apart from some left-over-kexts from previous iscsi initiator installs, which I now have removed. 

My config on the linux end:
-------------------------------------------------------------------------
    MaxRecvDataSegmentLength 8192
    MaxXmitDataSegmentLength 8192
    HeaderDigest None
    DataDigest None
    InitialR2T Yes
    MaxOutstandingR2T 1
    ImmediateData Yes
    FirstBurstLength 65536
    MaxBurstLength 262144
    DataPDUInOrder Yes
    DataSequenceInOrder Yes
    ErrorRecoveryLevel 0
    IFMarker No
    OFMarker No
    DefaultTime2Wait 2
    DefaultTime2Retain 20
    OFMarkInt Reject
    IFMarkInt Reject
    MaxConnections 10


<target iqn.2017-01.private.kroger.nas:target1>
	backing-store /dev/md2
	initiator-address 192.168.1.2
</target>
--------------------------------------------------------------------------

I will update the case as soon as I have any news. 

Kind regards, 
Søren

**Comment by sorenk** [comment]
For now I can stop the crashes by disabling icloud photos library, which is trying to download ~150GB of my orginial phots from icloud onto my iscsi disk. 

In the meanwhile, listening to iTunes which has the library on the same disk, doesn't cause any crashes (for several hours now). 
The disk has been mounted for nearly 24 hours now without crashes. 

So it seems it's related to (heavy?) writes or os x photos library. 
I have rsynced about 4GB to the iscsi disk without crashes. Will do some more write tests later on...

**Comment by nsinenian** [comment]
I've analyzed the crash log. It looks like the kernel panics when the iSCSI initiator tries to activate a connection. This is strange because it implies that there was no connection to begin with and you say this happens during large transfers. I'm wondering whether your connection is dropped and re-established during these large transfers. Keep me posted on your testing.

**Comment by TychoAI** [comment]
I just want to chime in to confirm I'm having a similar problem. It's similar, not identical, because I'm using it on a Mac with macOS Sierra, and the Synology DS216j NAS. The latest beta resolve my original issue of not being able to mount and format the target, as well as re-mounting after waking from sleep. It works well with small to medium size files, but when I try to move an approximately 150GB iPhoto library from the main drive to the iSCSI target, I get the same hard crash and reboot describe. It really only does seem to happen during large file transfers.

The main reason for trying to use this feature was to be able to offload the library to the NAS, so I've been sort of waiting in the wings to see if others are experiencing the same troubles.

**Comment by nsinenian** [comment]
@TychoAI can you post a crash log as well -- it would help with debugging to see if your log shows a similar backtrace. Can you also repeat the transfer, look at the memory pressure in Activity Monitor and tell me if that grows with the transfer? When does the crash occur into the transfer? (e.g., near the beginning, after many GBs have transferred, or ... ?)

@sorenk can you please post additional crash logs should they occur, so that I can check for consistency?

**Comment by sorenk** [comment]
Some more observations:
1) The console app contains lots of crash dumps from around the same time were the I get the kernel panics. I have attached those (there has been 2 kernel panics in the same timeframe). 
[Archive.zip](https://github.com/iscsi-osx/iSCSIInitiator/files/797651/Archive.zip)
2) Up to the first crash, I tried to capture some system information (like netstat, top, iostat, ping)
There was nothing scary to see. netstat reports the connection is established the whole time and performance seems to be stable all the way. There's no indication of "something building up" towards the kernel panic. 
3) I discovered that there are a lot of "stuck" processes on my mac mini in general compared to my macbook. From my understanding these are like uninterruptable sleeps which could be waiting for I/O. 
However, even though the number seems high to me (compared to my macbook which tops at 1 stuck process at a time), the number looks stable between 10-20 towards the kernel panik. Right now with icloud photosharing downloads set on pause the amount of stuck processes is about 10. 
Also, the processes which are stuck aren't all related to the iscsi disk (i think):
`home:~ root# ps aux|grep " U"
root             5203   0.0  0.1  2471648   4336 s002  U     8:58PM   0:00.03 sudo -i
kroeger          5121   0.0  0.2  2514644  18572   ??  Us    8:56PM   0:00.18 /System/Library/PrivateFrameworks/PhotoLibraryPrivate.framework/Versions/A/Frameworks/MediaConversionService.framework/Versions/A/XPCServices/com.apple.photos.VideoConversionService.xpc/Contents/MacOS/com.apple.photos.VideoConversionService
kroeger          5097   0.0  0.3  2518856  25664   ??  Us    8:56PM   1:22.28 /System/Library/PrivateFrameworks/PhotoLibraryPrivate.framework/Versions/A/Frameworks/MediaConversionService.framework/Versions/A/XPCServices/com.apple.photos.ImageConversionService.xpc/Contents/MacOS/com.apple.photos.ImageConversionService
kroeger          5037   0.0  0.7  2540904  55588   ??  U     8:55PM   1:35.59 /System/Library/PrivateFrameworks/PhotoLibraryPrivate.framework/Versions/A/Support/photolibraryd
kroeger          4963   0.0  0.2  2511460  13580   ??  U     8:55PM   0:00.13 /System/Library/CoreServices/iconservicesagent
_softwareupdate   627   0.0  0.8  2546896  64440   ??  Ss    3:51PM   0:01.48 /System/Library/CoreServices/Software Update.app/Contents/Resources/softwareupdated
_netbios          562   0.0  0.1  2473672   8312   ??  UNs   3:51PM   0:02.26 /usr/sbin/netbiosd
_devicemgr        452   0.0  0.1  2477480   8056   ??  Us    3:51PM   0:10.19 /Applications/Server.app/Contents/ServerRoot/System/Library/PrivateFrameworks/DeviceManagerModels.framework/XPCServices/dmpgHelper.xpc/Contents/MacOS/dmpgHelper
root              240   0.0  0.1  2498172   9708   ??  Us    3:51PM   0:01.43 /Applications/Server.app/Contents/ServerRoot/usr/libexec/servermetricsd -daemonize YES
root              216   0.0  0.1  2499088   7428   ??  Us    3:51PM   0:02.42 /sbin/emond
root              200   0.0  0.0  2472992   3668   ??  Us    3:51PM   0:00.07 /System/Library/Frameworks/Security.framework/Versions/A/XPCServices/com.apple.CodeSigningHelper.xpc/Contents/MacOS/com.apple.CodeSigningHelper
root               95   0.0  0.1  2498156   7232   ??  Us    3:51PM   0:01.05 /System/Library/CoreServices/launchservicesd
root               79   0.0  0.4  2632536  36920   ??  Us    3:51PM   0:53.98 /System/Library/Frameworks/CoreServices.framework/Frameworks/Metadata.framework/Support/mds
root             6022   0.0  0.0  2434840    772 s002  S+    9:42PM   0:00.00 grep  U`
4) I tried to hammer reproduce the issue with dd and writing various files in 4 simultaneous threads. At some point the dd processes became unresponsive. At the end, I wasn't able to get the system to kernel panic, but instead the disk got umounted / logged off. When I tried to logon again, I couldn't (I think i got a connection error) - when I tried to query the disk locally from the linux box I couldn't get a connection either. So I killed tgtd and restarted it. Afterwards I could logon from mac os again and mount the disk. However, the previous dd processes were still hanging (I suppose this is normal, after the iscsi target has crashed).
The only message i got from the tgtd process was:
`Feb 22 23:11:46 nas tgtd: tgtd: iscsi_task_rx_start(1794) Unknown op 2a`
5) Normally I don't see anything in the logs on the linux box when os x is crashing. And usually the tgtd process doesn't need to be restarted. However - today at the second crash, I got:
`Feb 23 15:51:19 nas tgtd: tgtd: conn_close(140) Forcing release of tx task 0x205ad10 3659d720 1`

I think I will have to try a newer version of tgtd.

**Comment by TychoAI** [comment]
nsinenian, I'm attaching the crash report. I tried it again last night. Copying a 130GB Photos library. It looks like it crashed about 1.5 hours after starting the copy. It looks like it got through about 15GBs of transfer. The only thing I can't really do is tell you about the memory usage. I can't sit there for hours and watch it heh...

The computer isn't set to go to sleep. It's connected to the NAS hardwired, on a Gigabit network. I hope that helps.

[iSCSI large copy crash.txt](https://github.com/iscsi-osx/iSCSIInitiator/files/801476/iSCSI.large.copy.crash.txt)

**Comment by nsinenian** [comment]
Thanks for the diagnostic information. It looks like you're both experiencing the same issue. The crash logs for iscsid appear to be more of an effect rather than a cause. I am wondering, during these large transfers does your machine go to sleep? Can you change power settings to never sleep and reproduce this problem?

**Comment by TychoAI** [comment]
The laptop is not set to go sleep, nor is it set to put drives to sleep (the main drive is SSD) so no spin down...the fact that it goes for well over an hour before crashing tells me it's not really related to sleep. So the log I sent is without setting the laptop to sleep.

The only thing that might have something to do with it (and this is just a shot in the dark) is that Photos libraries are a lot of hard links (I don't know the exact term) and the OS is supposed to handle copying actual files when you move the libraries. I don't know if that could have anything to do with it, but I thought I'd mention seeing how the other user's issue is also related to Photos.

**Comment by nsinenian** [comment]
@TychoAI Thanks - I missed your comment regarding sleep earlier. I'm not sure what you mean by "hard link" (symbolic links?). Filesystem-level objects should be completely transparent to iSCSI - it doesn't care what you're transferring or how, it just processes low-level hard disk commands. The logs show that the kernel extension crashes when the destructor for a component is called. This occurred when the connection is being taken down. That's why I asked about sleep, and earlier about whether the connection is interrupted at all. I will try to reproduce this.

**Comment by TychoAI** [comment]
I've tried to mitigate disconnection at the network level as much as I could, although there's still a possible point of failure at the switch. The laptop has a bettery, obviously, and NAS is on a UPS, but the switch gear could still be affected by some power fluctuation. I should probably get a small UPS for that too.

As for the Photos library, I'm pretty sure they are "hard links", not symbolic links (I had to look it up, as I couldn't remember the details). Looking forward to what you find. If I do get a UPS for the switch gear, I'll do another test.

**Comment by sorenk** [comment]
Actually, my crashes have disappeared since i disabled all power management / sleep related settings. 
I discovered it, when I had the photos iCloud download paused and the iscsi target was logged off the next day (the mac mini was setup to sleep during the night).

**Comment by oceanscape** [comment]
I'm having the same issue. I'm running iSCSIInitiator-1.0.0-beta5 on my Mac Pro (Early 2009) with El Capitan 10.11.3.

I have a 27 TB Drobo B800i attached. It's used exclusively as a Time Machine backup for all my other drives. The Drobo drive randomly disconnects during large file transfers (usually about a half hour or so into a backup).

**Comment by nsinenian** [comment]
@oceanscape Does your machine (kernel) crash during the disconnects? Is your machine set to go to sleep?

**Comment by oceanscape** [comment]
No to both.

Jeremy

**Comment by TychoAI** [comment]
I have since installed a UPS on the switching equipment and attempted the same large copy twice. In each instance the OS no longer crashed/panicked, but the copy still did stop after about 30GBs. I've not had time to dig up a crash report, but the error the Finder gives is that it was unable to to unmount the drive correctly which is weird as it didn't need unmounting, obviously. I'll attempt it again to try and capture the error and crash log and see if a 3rd time gives the same consistent result.

**Comment by nsinenian** [comment]
It would appear that whether it crashes or not seems to be related to power/sleep settings. However, the underlying problem is there regardless, with a dropped connection (and hence Finder’s complaint that it didn’t unmount properly — its as if the disk was unplugged). Since it appears to be repeatable, getting some good Wireshark capture of the iSCSI traffic around the time when the error occurs would go a long way to help trouble shoot.

Thanks.

**Comment by TychoAI** [comment]
here's what the errors are when it fails...not sure if I can set up/know how to do a WireShark capture, but I'll try. Can you filter while capturing and if so, what should my filter be?

![errors](https://cloud.githubusercontent.com/assets/22603085/24325474/d50999e6-1156-11e7-823e-fb8d15e96ab0.png)

**Comment by nsinenian** [comment]
You can filter the capture, and your filter should be "iSCSI".  If the issue is reproducible, I'd suggest that you capture traffic just before you think it's going to happen and until after it happens, or you will end up with a massive file. Thanks.  The screenshots you posted are typical when a connection is dropped - the question is why...

**Comment by TychoAI** [comment]
Until I can figure out how to do the capture, here are the crash logs associated with the last crash.

The problem isn't that I can't reproduce it...it crashes every time. The problem is I don't know how many hours into the copy it will crash. There's no away I can sit there waiting to know when it happens, so the only chance I have is to start the capture with the copy and hope for the best.

[Archive.zip](https://github.com/iscsi-osx/iSCSIInitiator/files/870595/Archive.zip)

**Comment by nsinenian** [comment]
@TychoAI Can you please try the new beta and let me know if it is any better?

The crash logs you posted helped, and I believe I have addressed that particular problem. However, the crash logs were for the daemon (iscsid), not the kernel extension. Unclear if they are related. In any case, this should move things forward. I look back to getting your feedback.

**Comment by TychoAI** [comment]
Hello Nareg,

thank you so much for getting back to this. I apologize for not revisiting this sooner. I will definitely give the new version a try and let you know the results, but I'll tell you a bit more of what happened with my last attempt, and why I haven't gotten back with a response yet.

After the last crash, there were another Sierra OS update. So I had finished that, uninstalled iSCSI and re-installed it. Then I turned off TimeMachine backups and re-started the large copy. This time the copy lasted much, much longer...about 100GBs in, but then it crashed in the same way. The Synology NAS I'm using is the TimeMachine target as well as the iSCSI target. Obviously, they are on separate partitions and volumes. I'm not sure if it has anything to do with it, but maybe that scenario helps with any other possible edge cases you'll encounter.

After all that, I was hoping to attempt the whole thing once again. Every single attempt so far has been using the MacBook's wireless interface. Every piece of network and hardware device is on UPS so I'm confident on the hardware, but the wireless is the last "magic" that could be an issue on long transfers. There are no wireless issues in general with the network. So I was hoping to borrow/buy a USB ethernet adapter, turn off TimeMachine, re-install iSCSI and try the whole thing again, but got bogged down with the kids and the fact that the wife (who's computer it is) can't seem be without her laptop long enough to run this test. heh

Anyway..I'll make this happen and let you know. Thanks again!

**Comment by TychoAI** [comment]
Success! Well...mostly.

I un-installed beta 5 and installed beta 6. I turned off TimeMachine and tried again. It crashed again after about 3 hours.

At this point the last thing I could try was to rule out the wireless connection. I borrowed a USB Ethernet adapter (it's a MacBook Air...has no wired Ethernet port) and plugged it in. Not only did this work, and copied over the entire 135GB library to the iSCSI target, but it did it in 32 minutes, rather than the projected 14 hours!

So the wired Ethernet helped in two ways. If there is a problem with iSCSIInitiator (I'm kind of thinking there isn't) it's definitely not size-based, but time based. I suspect this is unlikely. Having plugged in the wired Ethernet showed me that the MacBook wireless was going way slower than it should. It turns out that my router being set to serve mixed G and N causes the MacBook to connect at G speeds only. Since I no longer have the legacy devices that used G, I simply switched the router to N only. I tried doing the transfer again and this time the projected time was 7.5 hours instead of 14, so that's twice the speed. It's quite possible this would have worked, but I stopped it as I didn't want to mess around with it any further.

In short...it worked. Wired Ethernet dealt with the 135GB archive transfer without any troubles. The Photos library can be accessed like any other library in Photos for MacOS Sierra and it's pretty darn fast all things considered.

Many thanks, Nareg. Let me know if you need anything else. Limited time, but happy to help.
